### PR TITLE
Static PDF analysis changes

### PR DESCRIPTION
**Pull request by jgajek**
As discussed in issue #297.

1. Switch to packaged version of peepdf (thanks to @jbremer).

2. Don't call JSAnalyse() directly only on stream objects. Instead, rely on peepdf's getJavascriptCode() method, which extracts not just from stream objects, but also from strings and dictionaries.

3. Don't emulate Javascript code with PyV8 by default. Can enable emulation via emulate_js = yes in processing.conf.

4. Run PDF analysis in a separate process and impose a timeout (default 300 seconds).  Can adjust the timeout via pdf_timeout configuration variable in processing.conf.

### CODE REVIEW

**Comment by KillerInstinct** [comment]
Does this eliminate the need for PyV8 completely?

**Comment by jgajek** [comment]
If you leave the emulate_js config variable off, then yes.  If you turn it
on, then peepdf will import and call PyV8.

File: modules/processing/static.py [hunk 1]
```diff
@@ -954,16 +956,28 @@ def _parse(self, filepath):
...
+        p.start()
+        p.join(timeout=self.pdf_timeout)
+        if p.is_alive():
+            log.debug("PDF timeout exceeded, terminating analysis")
+            p.terminate()
+        try:
+            results = q.get(block=False)
+        except:
+            log.debug("No results from PDF analysis")
+            results = {"pdf":{}}
```
Comment (KillerInstinct) [inline modules/processing/static.py]: Should probably put some sort of warning/error here to be displayed in Django. Would help with collecting samples that are giving peepdf an issue.

Comment (KillerInstinct) [inline modules/processing/static.py]: 'here' = in the empty pdf dict. Didn't realize the comment would show 4 lines instead of 1 :)

**Comment by jgajek** [comment]
Any remaining concerns with this PR?

**Comment by mallorybobalice** [comment]
final hopeful bump ?

**Comment by mallorybobalice** [comment]
ps @jgajek - so unlike interrupting cow that sometimes fails for me as well when it freezes in DEBUG: About to parse with PDFParser 

this looks to also cater for that offense and kill the process it spawns to do the static pdf analysis, correct?

**Comment by jgajek** [comment]
@mallorybobalice I've been running this patch in a fairly busy SOC environment for a few weeks now without any major issues.  We did run into the occasional PDF analysis that times out after 5 minutes -- rudimentary profiling indicates that the majority of the time is spent in regular expression parsing of the PDF document.  Can't do much about it other than rewrite the parser, which is not something that I would  deem a priority at this point.

This is a fairly invasive PR -- perhaps it would be useful for others to try this patch in their environment to get more eyes on it.  It makes visible script-type objects (dictionaries, strings) which were previously not extracted in the analysis results.  There appear to be some issues with URL extraction within the peepdf code itself (i.e. URLs that should have been recognized and extracted, aren't) -- I haven't had time to dig into this yet.

**Comment by mallorybobalice** [comment]
@jgajek  - what are your thoughts on wrapping all of processing in similar timeouts not just static?
we've had a couple of issues where process.py worker processes stay there indefinitely after ES or Mongo crash and restart (related to memory management)... probably too much of a process.py rewrite and possibly unsafe I suppose.

**Comment by jgajek** [comment]
@mallorybobalice Do you mean the worker processes spawned from running this PR stay there indefinitely?

**Comment by mallorybobalice** [comment]
nope, PR is probably fine. the original process.py https://github.com/spender-sandbox/cuckoo-modified/issues/367 sorry for the confusion 

haven't had the change to merge your changes for our environment unfortunately. still hoping to see them merged first then do the usual updates

**Comment by mallorybobalice** [comment]
Closed as in won't be merged? Did Brad say why?

**Comment by jgajek** [comment]
@mallorybobalice Not to me directly.  Without a healthy feedback process, I'm going to be forced to fork unfortunately.

**Comment by mallorybobalice** [comment]
Any chance we could leave it open for a bit more for Brad to provide a closing comment? 
I think it's a worthwhile fork for cm stability

**Comment by spender-sandbox** [comment]
I'm working on fixing peepdf parsing issues actually, which IMO is a better solution than not ever having the bugs fixed due to what seems to be a dead upstream, and ending up with no static analysis at all (and wasting the max amount of CPU time each time a document of that kind is encountered).  Also fixing oletools bugs and submitting those upstream, will have more to commit soon, just doing some internal testing.

-Brad

**Comment by spender-sandbox** [comment]
As an overall comment though, my strategy for this repo involves actual understanding of problems and developing fixes based on that understanding.  If you want to paper over problems or disable functionality because some problem is too hard to understand, there's another repo I know of that might be more to your liking ;)  That's why people use this repo I think, because of it being based in actual malware analysis and solving the hard problems; it's not about flash or buzzwords.

-Brad

**Comment by mallorybobalice** [comment]
Brad, I think you misunderstand what we're saying. We love that you're writing a proper fix.

BUT

if CSB-M crashes in prod for people's SOC they need a workaround. They options are usually - ask the maintainer or write something on their knee . The latter is usually something done with great reluctance as any 'self written' fixes need to be diffed against upstream. 

jgajek PR provides a viable interim option UNTIL a 'proper fix' . 

In extreme cases I would bet people are considering turning off the PDF analysis module all together because of this but that's not awfully wise...(like we've turned off MHeur)

**Comment by jgajek** [comment]
I'm glad Brad found the time to look at the peepdf parsing code and hats off to him for finding that performance-hogging loop.  Had I known he was working on it in the short term, I wouldn't have bothered with the quick interim fix.  There are other features (that are more important to me) that I'm trying to find time for.  It's not a matter of something being "too hard to understand" -- it's a matter of prioritizing a very limited amount of time & resources.

**Pull request closed without merging**